✨ Add amp-empower component

merty opened the pull request:
We would like the users of our global ad network and video monetization platform, Empower, to be able to embed their videos hosted on our platform to their AMP pages, so we've created the `amp-empower` component.

merty commented:
This is the video player.

We opened a separate PR for the ad network integration: #24949

honeybadgerdontcare commented on extensions/amp-empower/validator-amp-empower.protoascii:
```diff
@@ -0,0 +1,45 @@
+    name: "amp-empower"
+    version: "0.1"
+    version: "latest"
+    requires_usage: GRANDFATHERED
```
please remove `requires_usage`

merty replied:
Removed 👍

rcebulko commented on extensions/amp-empower/OWNERS.yaml:
```diff
@@ -0,0 +1,5 @@
+# https://github.com/ampproject/amp-github-apps/blob/master/owners/OWNERS.example.yaml
+
+- ampproject/wg-ui-and-a11y
+- merty
```
Please use the JSON format for all new OWNERS files. This will become:

```javascript
// For an explanation of the OWNERS rules and syntax, see:
// https://github.com/ampproject/amp-github-apps/blob/master/owners/OWNERS.example

{
  rules: [
    {
      owners: [
        { name: "ampproject/wg-ui-and-a11y" },
        { name: "merty" }
      ]
    }
  ]
}
```

merty replied:
Done 👍

gmajoulet commented:
Adding @alanorozco since this is a video player :)

merty commented:
Rebased on `master`, resolved the conflict occured due to changes made in `extensions/amp-story/amp-story.md` and squashed the commits addressing the change requests.

gmajoulet commented:
ping @alanorozco for the video specific code

merty commented:
@alanorozco As we have users currently waiting for this component, we would really appreciate it if it could make it to the next release cycle.

Thanks in advance!

alanorozco commented on extensions/amp-empower/0.1/amp-empower.js:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    devAssert(this.video_);
+    let src =
```
nit: Can we make this a `const baseSrc`, and then assign inline?

```js
return (this.videoIframeSrc_ = addParamsToUrl(baseSrc, params));
```

merty replied:
Done 👍

alanorozco commented on extensions/amp-empower/0.1/amp-empower.js:
```diff
@@ -0,0 +1,114 @@
+
+  /** @override */
+  buildCallback() {
+    this.video_ = userAssert(
```
I suggest renaming to `videoId_` or similar. It's otherwise confused with actual `<video>` elements.

merty replied:
Done 👍

alanorozco commented on extensions/amp-empower/0.1/amp-empower.js:
```diff
@@ -0,0 +1,114 @@
+
+  /** @override */
+  layoutCallback() {
+    const iframe = this.element.ownerDocument.createElement('iframe');
```
Use `createFrameFor` from `src/iframe-video.js`. It has a standard interface to accomplish what you're doing here.

merty replied:
Done 👍

merty commented:
@alanorozco Honestly, I wasn't aware of them, thanks for the heads-up!

Just rewrote the component to implement the `VideoInterface` and register to the `VideoManager`. It now supports actions, autoplay, docking and rotate-to-fullscreen.

Could you please re-review it?

merty commented:
@alanorozco I would like to kindly remind you of this PR.

Thank you for your time!

merty commented:
@alanorozco Once again, I would like to kindly remind you of this PR.

alanorozco commented:
@merty Could you please explain why `amp-video-iframe` doesn't fulfill this use case?

merty commented:
@alanorozco The same reason why other AMP video components exist: In order to provide the non-tech-savvy users of our service a straightforward way to integrate our video player on their AMP pages and spare them dealing with issues due to misconfiguration.

I believe this is a valid use case considering that the AMP documentation suggests using the `amp-video-iframe` component _if_ the platform does not have an AMP component.

From the [amp-video-iframe documentation](https://amp.dev/documentation/components/amp-video-iframe/) (emphasis mine):

> When should I use this?
> 
> This component is useful if you've built your own Javascript-based video player and would like to embed it in an AMP document, **or if your player is provided by a third party not supported by the AMP component library**.
> 
> 1. If you'd like to include a video directly on the AMP document, you should use amp-video.
> 
> 2. **If you're using a common 3rd party like Youtube, Vimeo or others supported in AMP, you should use their supported component (e.g. amp-youtube, amp-vimeo).**
> 
> 3. If you've built a custom player or are using one provided by an unsupported 3rd party, you should use amp-video-iframe. This is different from using amp-iframe in that it enables Video Features on AMP. See behavior below for more details.

alanorozco commented:
> The same reason why other AMP video components exist: In order to provide the non-tech-savvy users of our service a straightforward way to integrate our video player on their AMP pages and spare them dealing with issues due to misconfiguration.

That's one benefit, for sure, but it's not the sole reason. Historically, `amp-video-iframe` was not available at the time most of these components were built. All of these have a maintenance cost even if no features are added, since AMP is an evergreen project and we continuously update our codebase in general sweeps even if a specific component doesn't change.

Regarding the snippet of documentation you copied as justification for this component, I'm afraid that does not apply. The document you copied is meant to be read for user's, so that's explained from the point of view of what may be available to them. It does not explain when a _new_ player component should be contributed.

To fill this gap, there's currently #25287 where we're introducing a general explainer for cases where a component should be created or `amp-video-iframe` should be used instead.

Configuration differences between this component and `amp-video-iframe` would be negligible as long as you provide a harness document in which authors can simply drop in their videos referenced by URL. This is not much different than configuring attributes.

If there aren't any specific features that this player supports that `amp-video-iframe` does not, then I will keep this issue closed.

merty commented:
Thank you for not taking 5 minutes to respond for **29 days** and rejecting due to a change of contribution policy that's not even public yet.

I wish you simply rejected **1.5 months** ago, so we wouldn't waste our time waiting.

alanorozco commented:
@merty It's not a change in policy. Video players are accepted on a discretionary basis, and you did not follow general AMP contribution process by opening an I2I issue first so this was not able to be discussed properly.

I apologize for the delay but I'll have you know that reviewing your component is not my sole responsibility. I understand that it's frustrating to not get your change in, but you didn't take the proper steps.
